### PR TITLE
ios support

### PR DESCRIPTION
**Pull request by jet082**
Tested and working.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -615,10 +616,21 @@ else ifneq (,$(findstring ios,$(platform)))
...
-	CC = clang -arch armv7 -isysroot $(IOSSDK)
+	ifeq ($(platform), ios-arm64)
+		NO_REC = 1
+		CC = clang -arch arm64 -isysroot $(IOSSDK)
+		CXX = clang++ -arch arm64 -isysroot $(IOSSDK)
+		CXXFLAGS += -DIOS -DHAVE_OPENGLES3
+		CFLAGS += -Wno-error=implicit-function-declaration -DIOS -DHAVE_OPENGLES3
+		HAVE_GENERIC_JIT := 0
+		WITH_DYNAREC = DYNAREC_NONE
+		HOST_CPU_FLAGS = -DHOST_CPU=$(DYNAREC_NONE)
```
Comment (flyinghead) [inline Makefile]: This is not the way `HOST_CPU` should be used. For iOS it's supposed to be HOST_CPU_ARM64. You can add an iOS target in build.h and set the various defines there (`FEAT_SHREC`, `FEAT_AREC`, `FEAT_DSPREC`, ...)

File: core/hw/sh4/sh4_core_regs.cpp [hunk 2]
```diff
@@ -112,7 +112,7 @@ void SetFloatStatusReg(void)
             : "=r"(raa)
             : "r"(x), "r"(y)
            );
-#elif HOST_CPU == CPU_ARM64
+#elif HOST_CPU == CPU_ARM64 || defined(IOS)
```
Comment (flyinghead) [inline core/hw/sh4/sh4_core_regs.cpp]: This shouldn't be needed if `HOST_CPU` is correct

File: core/libretro/vmem_utils.cpp [hunk 3]
```diff
@@ -47,7 +47,9 @@ bool mem_region_lock(void *start, size_t len)
 {
 	size_t inpage = (uintptr_t)start & PAGE_MASK;
 	if (mprotect((u8*)start - inpage, len + inpage, PROT_READ))
-		die("mprotect failed...");
+		#ifndef IOS
```
Comment (flyinghead) [inline core/libretro/vmem_utils.cpp]: I don't think it's OK if this call fails on iOS. Probably it shouldn't be called at all, or the result should be checked.
This is required to detect texture updates even when using the interpreter.

File: core/libretro/vmem_utils.cpp [hunk 4]
```diff
@@ -56,15 +58,19 @@ bool mem_region_unlock(void *start, size_t len)
 	size_t inpage = (uintptr_t)start & PAGE_MASK;
 	if (mprotect((u8*)start - inpage, len + inpage, PROT_READ | PROT_WRITE))
 		// Add some way to see why it failed? gdb> info proc mappings
-		die("mprotect  failed...");
+		#ifndef IOS
```
Comment (flyinghead) [inline core/libretro/vmem_utils.cpp]: same comment as above

File: core/libretro/vmem_utils.cpp [hunk 5]
```diff
@@ -56,15 +58,19 @@ bool mem_region_unlock(void *start, size_t len)
...
+		#endif
 	return true;
 }
 
 bool mem_region_set_exec(void *start, size_t len)
 {
 	size_t inpage = (uintptr_t)start & PAGE_MASK;
 	if (mprotect((u8*)start - inpage, len + inpage, PROT_READ | PROT_WRITE | PROT_EXEC))
-		die("mprotect  failed...");
+		#ifndef IOS
```
Comment (flyinghead) [inline core/libretro/vmem_utils.cpp]: Same although this one is only used when using a dynarec iirc

File: core/serialize.cpp [hunk 6]
```diff
@@ -1127,7 +1127,9 @@ bool dc_serialize(void **data, unsigned int *total_size)
 	LIBRETRO_S(reg_dimm_status);
 	LIBRETRO_S(NaomiDataRead);
 
-	LIBRETRO_S(cycle_counter);
+	#if FEAT_SHREC != DYNAREC_NONE
```
Comment (flyinghead) [inline core/serialize.cpp]: This makes the save state format differ when using a dynarec and when not using one. Here the `cycle_counter` doesn't need to be de/serialized anyway so just do:
```
    LIBRETRO_S(i); // cycle_counter
```

File: core/serialize.cpp [hunk 7]
```diff
@@ -1628,7 +1630,9 @@ bool dc_unserialize(void **data, unsigned int *total_size, size_t actual_data_si
 	   LIBRETRO_US(dummy_int);		// NAOMI_COMM_DATA
 	}
 
-	LIBRETRO_US(cycle_counter);
+	#if FEAT_SHREC != DYNAREC_NONE
```
Comment (flyinghead) [inline core/serialize.cpp]: Same problem as above. Simply do:
```
    LIBRETRO_US(dummy_int); // cycle_counter
```

**Review by flyinghead — CHANGES_REQUESTED** [review]

**Comment by inactive123** [comment]
Hi @jet082, can you rebase this?

**Comment by dankcushions** [comment]
closed during hack-gate 2020. @twinaphex please re-open :)

**Pull request closed without merging**